// @sanity/base/lib/components/Document.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _AppLoadingScreen = require('./AppLoadingScreen');

var _AppLoadingScreen2 = _interopRequireDefault(_AppLoadingScreen);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function assetUrl(staticPath, item) {
  var isAbsolute = item.path.match(/^https?:\/\//);
  if (isAbsolute) {
    return item.path;
  }

  var base = staticPath + '/' + item.path;
  if (!item.hash) {
    return base;
  }

  var hasQuery = base.indexOf('?') !== -1;
  var separator = hasQuery ? '&' : '?';
  return '' + base + separator + item.hash;
}

function Document(props) {
  var stylesheets = props.stylesheets.map(function (item) {
    return _react2.default.createElement('link', {
      key: item.path,
      rel: 'stylesheet',
      href: assetUrl(props.staticPath, item)
    });
  });

  var scripts = props.scripts.map(function (item) {
    return _react2.default.createElement('script', {
      key: item.path,
      src: assetUrl(props.staticPath, item)
    });
  });

  var favicons = props.favicons.map(function (item, index) {
    return _react2.default.createElement('link', {
      key: item.path + index,
      rel: 'icon',
      href: assetUrl(props.staticPath, item)
    });
  });

  return _react2.default.createElement(
    'html',
    null,
    _react2.default.createElement(
      'head',
      null,
      _react2.default.createElement('meta', { charSet: 'utf-8' }),
      _react2.default.createElement(
        'title',
        null,
        props.title
      ),
      _react2.default.createElement('link', { rel: 'stylesheet', type: 'text/css', href: 'https://cdnjs.cloudflare.com/ajax/libs/normalize/5.0.0/normalize.min.css' }),
      stylesheets,
      favicons,
      _react2.default.createElement('meta', { name: 'viewport', content: props.viewport })
    ),
    _react2.default.createElement(
      'body',
      null,
      _react2.default.createElement(
        'div',
        { id: 'sanity' },
        _react2.default.createElement(_AppLoadingScreen2.default, { text: props.loading })
      ),
      scripts
    )
  );
}

var asset = _react.PropTypes.shape({
  path: _react.PropTypes.string.isRequired,
  hash: _react.PropTypes.string
});

Document.defaultProps = {
  charset: 'utf-8',
  title: 'Sanity',
  viewport: 'width=device-width, initial-scale=1',
  loading: 'Loading Sanity…',
  staticPath: '/static',
  favicons: [{ path: 'favicon.ico' }],
  stylesheets: [],
  scripts: []
};

Document.propTypes = {
  charset: _react.PropTypes.string,
  title: _react.PropTypes.string,
  viewport: _react.PropTypes.string,
  loading: _react.PropTypes.node,
  staticPath: _react.PropTypes.string,
  favicons: _react.PropTypes.arrayOf(asset),
  stylesheets: _react.PropTypes.arrayOf(asset),
  scripts: _react.PropTypes.arrayOf(asset)
};

exports.default = Document;
