// @sanity/base/lib/components/buttons/DefaultButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// import Ink from 'react-inc'
// import styles from 'part:@sanity/base/theme/buttons/default-style'

var DefaultButton = function (_React$Component) {
  _inherits(DefaultButton, _React$Component);

  function DefaultButton() {
    _classCallCheck(this, DefaultButton);

    return _possibleConstructorReturn(this, (DefaultButton.__proto__ || Object.getPrototypeOf(DefaultButton)).apply(this, arguments));
  }

  _createClass(DefaultButton, [{
    key: "render",
    value: function render() {
      var _props = this.props,
          onClick = _props.onClick,
          kind = _props.kind;


      return _react2.default.createElement(
        "button",
        {
          type: "button",
          title: "Clear value",
          onClick: onClick
        },
        _react2.default.createElement(
          "span",
          null,
          this.props.icon
        ),
        _react2.default.createElement(
          "span",
          null,
          this.props.children
        )
      );
    }
  }]);

  return DefaultButton;
}(_react2.default.Component);

DefaultButton.propTypes = {
  kind: _react.PropTypes.string,
  onClick: _react.PropTypes.func,
  children: _react.PropTypes.node,
  inverted: _react.PropTypes.bool,
  icon: _react.PropTypes.node
};
exports.default = DefaultButton;
