// @sanity/base/lib/datastores/utils/createActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createActions;

var _pubsubber = require('./pubsubber');

var _pubsubber2 = _interopRequireDefault(_pubsubber);

var _SanityStoreObservable = require('./SanityStoreObservable');

var _SanityStoreObservable2 = _interopRequireDefault(_SanityStoreObservable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Todo: refactor out
function createActions(actions) {
  return Object.keys(actions).reduce(function (acc, name) {
    acc[name] = createAction(name, actions[name]);
    return acc;
  }, {});
}

function createAction(name, fn) {

  var calls = (0, _pubsubber2.default)();

  var functor = function action() {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var retValue = fn.apply(undefined, args);

    calls.publish({
      name: name,
      progress: retValue.progress,
      returnValue: retValue,
      args: args
    });

    return retValue;
  };

  functor.calls = new _SanityStoreObservable2.default(function (observer) {
    return calls.subscribe(function (call) {
      observer.next(call);
    });
  });

  return functor;
}
